lib/Backends/CPU/LLVMIRGen.cpp

```diff
@@ -676,6 +655,50 @@ void LLVMIRGen::emitDataParallelKernel(
   createCall(builder, kernelFunc, buffers);
 }
 
+/// Emit the function that implements a data-parallel kernel and calls it.
+///
+/// The generated kernel functions get buffers as their parameters. The buffers
+/// are uniqued, so that any buffer is passed as argument to the kernel function
+/// only once. This allows us to mark all parameters of the generated kernel as
+/// noalias. As a result, the LLVM optimizer makes use of the noalias attributes
+/// and produces nicely vectorized code for the generated data-parallel kernels.
+void LLVMIRGen::emitDataParallelKernel(
+    llvm::IRBuilder<> &builder, llvm::ArrayRef<const Instruction *> bundle) {
+  if (bundle.empty())
+    return;
+  // Types of arguments for the kernel function being generated.
+  llvm::SmallVector<llvm::Type *, 32> argTypes;
+  // Map each buffer used by the kernel to the argument number of the kernel
+  // function. This ensures that same buffer is always mapped to the same
+  // argument.
+  llvm::DenseMap<Value *, int> bufferToArgNum;
+  // Buffers to be passed to the kernel function as arguments.
+  llvm::SmallVector<llvm::Value *, 32> buffers;
+  // Collect unique buffers used by the instructions of the kernel.
+  llvm::SmallVector<const Instruction *, 32> batchedBundle;
+  for (const auto I : bundle) {
+    if (argTypes.size() + I->getOperands().size() > 64) {
+      emitDataParallelKernelImpl(builder, batchedBundle, argTypes,
+                                 bufferToArgNum, buffers);
+      batchedBundle.clear();
+      argTypes.clear();
+      bufferToArgNum.clear();
+      buffers.clear();
+    }
+    batchedBundle.push_back(I);
+    for (const auto &Op : I->getOperands()) {
```

Comment:
In theory, to be very exact, we should look how many unique operands a given instruction has and then check if `argTypes.size() + numUniqueOperands > 64`. Your check above `argTypes.size() + I->getOperands().size() > 64` is correct, but more pessimistic.

Reply:
Yes, it is perssimistic, I forgot to add this comments. But I think the trade-off is that the algorithm becomes simpler and it the downside isn't much. So algorithm wise, I think it is good.